### PR TITLE
🏃 allow extra args with tilt

### PR DESCRIPTION
**Pull request by alexeldeib**
Allows extra args from the tilt provider config.

This makes testing things like feature flags much easier. It needs to be in provider config rather than settings so it can be used across all providers (settings files are 1 per invocation I think? but we could change that).

### CODE REVIEW

File: Tiltfile [hunk 1]
```diff
@@ -25,21 +25,6 @@ default_registry(settings.get("default_registry"))
...
-            "go.sum",
-            "api",
-            "cmd",
-            "controllers",
-            "errors",
-            "third_party",
-            "util",
-        ],
-    },
     "kubeadm-bootstrap": {
```
Comment (detiber) [inline Tiltfile:28]: Wouldn't we also need to pass in the feature gate to the `kubeadm-bootstrap` provider as well?

File: docs/book/src/developer/tilt.md [hunk 2]
```diff
@@ -192,6 +192,29 @@ COPY --from=tilt-helper /usr/bin/docker /usr/bin/docker
...
+
+Example:
+
+```json
+{
+    "name": "core",
+    "config": {
+        "context": ".",
+        "image": "gcr.io/k8s-staging-cluster-api/cluster-api-controller",
+        "extra_args": ["--feature-gates=MachinePool=true"]
```
Comment (detiber) [inline docs/book/src/developer/tilt.md:207]: If I'm understanding this correct, in order to test a feature gate, you would have to override the necessary `tilt-provider.json` files with the correct `extra_args`. Is that correct?

I'm wondering if we can simplify the developer configuration a bit by giving a mapping of provider -> extra_args in `tilt-settings.json` instead?

Comment (alexeldeib) [reply docs/book/src/developer/tilt.md:207]: I'm a little unfamiliar with tilt -- is tilt-settings just an "us" thing and there's not any particular convention around using it? If so I definitely agree.

I waffled back and forth on this, it wasn't clear to me if tilt-settings was supposed to be like a driver for the specific invocation, or if it's more just a lib and we can do whatever we want. Sounds like we can go for it?

File: Tiltfile (hunk 1 above)
Comment (alexeldeib) [reply Tiltfile:28]: yep, if this is agreeable overall I can shuffle those over too. Sounds like we can probably pull from tilt-settings and I'll revert the tilt-provider changes

File: docs/book/src/developer/tilt.md (hunk 2 above)
Comment (detiber) [reply docs/book/src/developer/tilt.md:207]: It's an "us" thing, we created it as a way to add config for things like which downstream repos to consume and injecting things like the base64 encoded credentials for providers, etc

File: docs/book/src/developer/tilt.md [hunk 3]
```diff
@@ -192,6 +192,29 @@ COPY --from=tilt-helper /usr/bin/docker /usr/bin/docker
 COPY --from=tilt-helper /go/kubernetes/client/bin/kubectl /usr/bin/kubectl
 ```
 
+
+**extra_args** (Array, default=[]): An array of additional arguments to pass to the main binary configured
+for this provider. Each item in the array will be passed is to the manager.
```
Comment (detiber) [inline docs/book/src/developer/tilt.md:197]:
```suggestion
for this provider. Each item in the array will be passed in to the manager. 
```

File: Tiltfile [hunk 4]
```diff
@@ -156,14 +159,17 @@ def enable_provider(name):
         additional_docker_build_commands,
     ])
 
-    # Set up an image build for the provider. The live update configuration syncs the output from the local_resource
-    # build into the container.
```
Comment (detiber) [inline Tiltfile:160]: Should we keep this comment?

File: Tiltfile [hunk 5]
```diff
@@ -95,7 +95,10 @@ COPY --from=tilt-helper /go/kubernetes/client/bin/kubectl /usr/bin/kubectl
 #     }
 # }
 def load_provider_tiltfiles():
-    provider_repos = settings.get("provider_repos", [])
+    provider_repos = settings.get("provider_repos", ["."])
+    if "." not in provider_repos:
+        provider_repos.append(".")
```
Comment (detiber) [inline Tiltfile:100]: Is this still needed with extra_args going in tilt-settings.json now?
```suggestion
    provider_repos = settings.get("provider_repos", [])
```

File: docs/book/src/developer/tilt.md [hunk 6]
```diff
@@ -192,6 +192,29 @@ COPY --from=tilt-helper /usr/bin/docker /usr/bin/docker
 COPY --from=tilt-helper /go/kubernetes/client/bin/kubectl /usr/bin/kubectl
 ```
 
+
+**extra_args** (Array, default=[]): An array of additional arguments to pass to the main binary configured
```
Comment (detiber) [inline docs/book/src/developer/tilt.md:196]:
```suggestion
**extra_args** (Object, default={}): A mapping of provider to additional arguments to pass to the main binary configured
```

File: docs/book/src/developer/tilt.md [hunk 7]
```diff
@@ -192,6 +192,29 @@ COPY --from=tilt-helper /usr/bin/docker /usr/bin/docker
...
+Example:
+
+```json
+{
+    "name": "core",
+    "config": {
+        "context": ".",
+        "image": "gcr.io/k8s-staging-cluster-api/cluster-api-controller",
+        "extra_args": ["--feature-gates=MachinePool=true"]
+    }
```
Comment (detiber) [inline docs/book/src/developer/tilt.md:208]: Unless I'm mistaken, the example should look more like this as part of tilt-settings.json, it currently looks like it is still formatted for the previous implementation.
```suggestion
    "extra_args": {
        "core": ["--feature-gates=MachinePool=true"],
        "kubeadm-bootstrap": ["--feature-gates=MachinePool=true"],
        "azure": ["--feature-gates=MachinePool=true"]
    }
```

File: tilt-provider.json [hunk 8]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (detiber) [inline tilt-provider.json:1]: I don't believe this is needed anymore with the current implementation.

File: Tiltfile (hunk 5 above)
Comment (alexeldeib) [reply Tiltfile]: fixed 👍

File: docs/book/src/developer/tilt.md [hunk 9]
```diff
@@ -192,6 +192,28 @@ COPY --from=tilt-helper /usr/bin/docker /usr/bin/docker
...
+**extra_args** (Object, default={}): A mapping of provider to additional arguments to pass to the main binary configured
+for this provider. Each item in the array will be passed in to the manager. 
+
+Example:
+
+```json
+{
+    "extra_args": {
+        "core": ["--feature-gates=MachinePool=true"],
+        "kubeadm-bootstrap": ["--feature-gates=MachinePool=true"],
```
Comment (alexeldeib) [inline docs/book/src/developer/tilt.md:205]: does this actually do anything? I'm not using it AFAIK, but I don't recall if kubeadm-bootstrap specifically uses this value at all

Comment (detiber) [reply docs/book/src/developer/tilt.md:205]: It looks like it's used for adding the watch on MachinePool: https://github.com/kubernetes-sigs/cluster-api/blob/99a95aaa59492924b04e6dddbea17992fecca24d/bootstrap/kubeadm/controllers/kubeadmconfig_controller.go#L109-L116 and also in determining the config owner (which is used during the Reconcile method): https://github.com/kubernetes-sigs/cluster-api/blob/99a95aaa59492924b04e6dddbea17992fecca24d/bootstrap/util/configowner.go#L95-L100

Comment (alexeldeib) [reply docs/book/src/developer/tilt.md:205]: thanks for the links, probably should have tried a simple ctrl-f 😋

**Pull request merged**